### DISCUSSION TITLE
Wrong database connection in use when working with SwitchTenantDatabaseTask

### DISCUSSION THREAD

**Discussion by m7vm7v** [spatie/laravel-multitenancy#439] [Q&A]
Great package, it helps a lot!

It seems the database connection does not switch to the tenant's.

In my main layout, I did try to get the `{{ dd(\Illuminate\Support\Facades\DB::getConfig()) }}` and it points to the 'landlord', but i'd expect at this point to be the tenant's. If I do `{{ dd(app('currentTenant')->name) }}` I see the correct tenant. 

I have followed the video tutorial and set up my environment and is now identical.

If I update the `SwitchTenantDatabaseTask` with `DB::setDefaultConnection($tenantConnectionName);` at the end it starts working as intended but I am under the impression that should happen automatically through the package and indicating I might be missing a step.

Can you please advise if that is intended and if my approach is fine (well if that is the solution will make it a custom task and not from the vendor)? Any pointers are greatly appreciated.

**Comment by masterix21** [comment]
Why should you change the default connection from `landlord` to `tenant`?

**Reply by m7vm7v** [reply]
Hmmm, isn't that the intended way? I think I am doing something wrong then...
When I added the `DB::setDefaultConnection($tenantConnectionName);` it did start changing the tenant to the correct one. Without that line when I load the tenant url an exception is thrown about not finding a datatable that exists only on the tenant and not the landlord. Indicating that is picking the wrong database (the landlord, when at that stage I'd expect it to be the tenant).

**Reply by masterix21** [reply]
I don't know, but feel free to add me to your repository: I will help you to solve

**Reply by nick-nds** [reply]
@m7vm7v @masterix21  I am also facing same problem. I went through the SwitchTenantDatabaseTask code. I found out the closure to extend DatabaseManager is not being executed which switches the database connection.

Line 41 of SwitchTenantDatabaseTask .PHP
```
app('db')->extend($tenantConnectionName, function ($config, $name) use ($databaseName) {
  $config['database'] = $databaseName;
  Log::info("check connection");

  return app('db.factory')->make($config, $name);
});
```

Adding a Log does not log anything from the closure. Replacing DatabaseManager extending with `DB::setDefaultConnection($tenantConnectionName)` fixes the issue.

Also, I am using this package with Laravel 10.

**Comment by boguslaw** [comment]
Hi. I notice the same issue. I have fresh installation with Laravel 10.2.0, PHP 8.2.3, laravel-multitenancy 3.0.2. I checked manuals, and all looks good. 

I used laravel-multitenancy with Laravel 8.x and 9.x and was working perfektly :)

**Comment by iqbalmalik89** [comment]
Is there any fix?

**Reply by m7vm7v** [reply]
What you can do is create your own SwitchTenantDatabaseTask -

```php
use Illuminate\Support\Facades\DB;
use Spatie\Multitenancy\Tasks\SwitchTenantDatabaseTask as SpatieSwitchTenantDatabaseTask;

class SwitchTenantDatabaseTask extends SpatieSwitchTenantDatabaseTask
{
    protected function setTenantConnectionDatabaseName(?string $databaseName)
    {
        parent::setTenantConnectionDatabaseName($databaseName);

        $tenantConnectionName = is_null($databaseName)
            ? $this->landlordDatabaseConnectionName()
            : $this->tenantDatabaseConnectionName();

        DB::setDefaultConnection($tenantConnectionName);
    }
}
```

Then in the multitency config file replace the given task with yours.

**Reply by ADezon** [reply]
A mixture of both, is the correct solution. I was losing my mind with this issue (Laravel 10), but creating a custom class and declaring it at config/multitenancy.php resolves this issue!

'switch_tenant_tasks' => [
//        \Spatie\Multitenancy\Tasks\SwitchTenantDatabaseTask::class,
        \App\Actions\Tipster\TipsterSwitchTenantDatabase::class,

Of course, witn "DB::setDefaultConnection($tenantConnectionName);" in substitution of the whole "app('db')->extend($tenantConnectionName... etc.

that's all.
